Sandboxes: Fix generation

ndelangen opened the pull request:
## What I did

The sandbox generation script had been failing for a long time.

I'm making it work again, by addressing multiple underlaying issues.

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

<!-- Please include the steps to test your changes here. For example:

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access X story

-->

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

<!-- BENCHMARK_SECTION -->
<!-- BENCHMARK_SECTION -->

<!-- greptile_comment -->

## Greptile Summary

This PR fixes sandbox generation issues in the Storybook project by addressing problems with local registry handling and Yarn 1 compatibility. The main changes include:

**Sandbox Generation Script Refactoring**: The `scripts/sandbox/generate.ts` file has been significantly refactored to improve local registry handling. The `withLocalRegistry` function was simplified by removing its dependency on the package manager and hardcoding the registry URL restoration. A new `addResolutions` function was introduced that directly modifies `package.json` files instead of using the package manager API for adding resolutions. The main generation process is now wrapped in a `withLocalRegistry` context to ensure all sandbox operations occur within the proper registry environment.

**Import Path Optimization**: The `code/core/src/shared/status-store/index.ts` file was updated to use a more local import path for the `StoryId` type, changing from `'storybook/internal/csf'` to `'../../types'`. This improves the internal module structure while maintaining the same functionality since the types module re-exports the same type.

**Yarn 1 CI Compatibility**: The `Yarn1Proxy.ts` file was modified to conditionally include the `--ignore-workspace-root-check` flag based on the environment. In CI environments, this flag is omitted, while in local development it's included. This addresses installation failures that were occurring in CI environments during sandbox generation.

These changes work together to resolve sandbox generation failures that were likely caused by package manager API limitations, registry configuration issues, and environment-specific Yarn 1 behavior. The refactoring moves away from relying on package manager abstractions toward more direct file system operations for better reliability across different environments.

**PR Description Notes:**
- The "What I did" section is empty and should be filled with a description of the changes
- No testing approach is specified in the manual testing section
- No documentation updates are indicated despite the internal changes

## Confidence score: 4/5

- This PR addresses specific technical issues with sandbox generation and appears well-targeted to fix the underlying problems
- Score reflects solid technical approach but some complexity in the registry handling changes that could have edge cases
- Pay close attention to the `scripts/sandbox/generate.ts` file due to significant refactoring of registry and package resolution logic

<!-- /greptile_comment -->

ndelangen committed:
Refactor: Update import statements and enhance CI detection in sandbox script

- Changed import of StoryId to use the correct path in status-store.
- Modified CI detection logic to include both GITHUB_ACTIONS and CI environment variables.
- Simplified local registry handling in the sandbox generation script.
- Added console logs for better debugging during the Storybook initialization process.

ndelangen committed:
Enhance Yarn1Proxy and sandbox generation script

- Updated Yarn1Proxy to conditionally set install arguments based on CI environment.
- Added functionality to add resolutions in the sandbox generation script for better dependency management.
- Improved error handling and logging during sandbox generation process.

ndelangen committed:
Update GitHub Actions workflow for sandbox generation

- Changed cron syntax to use single quotes for consistency.
- Updated branch references in the workflow to target 'norbert/fix-sandbox-generation'.
- Adjusted failure notification message to reflect the correct branch context.
- Commented out the 'generate-main' job to prevent execution during this workflow.

ndelangen committed:
Fix package.json reading in sandbox generation script

- Updated the readFile function to specify 'utf-8' encoding when reading package.json.
- Changed the JSON parsing method to use JSON.parse for better clarity and consistency.

ndelangen committed:
Update SOLID generator options and clean up sandbox generation script

- Modified the SOLID generator to include an option to disable component addition.
- Removed commented-out resolution for 'jackspeak' from the sandbox generation script to streamline the code.

ndelangen committed:
reset

ndelangen commented on code/core/src/common/js-package-manager/Yarn1Proxy.ts:
```diff
@@ -36,7 +36,7 @@ export class Yarn1Proxy extends JsPackageManager {
 
   getInstallArgs(): string[] {
     if (!this.installArgs) {
-      this.installArgs = ['--ignore-workspace-root-check'];
```
This one is puzzling... but this flag seemed to not be allowed sometimes, even though it's yarn 1.

I do not have an explanation for why... but I do know the flag is never useful for sandbox we generate, because those are never generated in a monorepo, but always in a temp directory, in isolation.

ndelangen commented on scripts/sandbox/generate.ts:
```diff
@@ -104,27 +99,13 @@ const addStorybook = async ({
 
-    const packageManager = JsPackageManagerFactory.getPackageManager({ force: 'yarn1' }, tmpDir);
     if (localRegistry) {
-      await withLocalRegistry({
-        packageManager,
-        action: async () => {
-          await packageManager.addPackageResolutions({
-            ...storybookVersions,
-            // Yarn1 Issue: https://github.com/storybookjs/storybook/issues/22431
-            jackspeak: '2.1.1',
-          });
-
-          await sbInit(tmpDir, env, [...flags, '--package-manager=yarn1'], debug);
-        },
-        cwd: tmpDir,
-        env,
-        debug,
-      });
```
Instead of wrapping every sandbox generation with a `withLocalRegistry`, I wrapped the whole job in it, greatly simplifying and reducing the amount of noise in the job.

ndelangen commented on scripts/sandbox/generate.ts:
```diff
@@ -104,27 +99,13 @@ const addStorybook = async ({
   try {
     await copy(beforeDir, tmpDir);
 
-    const packageManager = JsPackageManagerFactory.getPackageManager({ force: 'yarn1' }, tmpDir);
```
This was the line, the CI got stuck on. 

Since it's used in a very limited way, it's not adding a lot to actually use it, so by replacing this with a direct in-file implementation the CI was no longer blocked.

ndelangen committed:
Enhance Yarn1Proxy tests with process environment mocking

- Added a mock for the 'node:process' module in Yarn1Proxy.test.ts to set the CI environment variable to 'false'.
- Updated Yarn1Proxy.ts to import the 'node:process' module for improved environment handling.

ndelangen committed:
Merge branch 'next' into norbert/fix-sandbox-generation

ndelangen committed:
Update Yarn1Proxy tests to set CI environment variable to 'false'

- Modified the mock for the 'node:process' module in Yarn1Proxy.test.ts to explicitly set the CI environment variable to 'false' for improved test reliability.

ndelangen committed:
Refactor Yarn1Proxy test environment mocking

- Updated the mock for the 'node:process' module in Yarn1Proxy.test.ts to set the CI environment variable to a boolean value instead of a string, enhancing consistency in test behavior.

yannbf commented on scripts/sandbox/generate.ts:
```diff
@@ -104,27 +97,13 @@ const addStorybook = async ({
+
+    await sbInit(tmpDir, env, [...flags, '--package-manager=yarn1'], debug);
   } catch (e) {
+    console.log('error', e);
```
is this intentional?

yannbf commented on scripts/sandbox/generate.ts:
```diff
@@ -397,7 +399,15 @@ if (esMain(import.meta.url)) {
     .option('--local-registry', 'Use local registry', false)
     .action((optionValues) => {
-      generate(optionValues)
+      withLocalRegistry({
```
This should only be called if `--local-registry` flag is passed

ndelangen committed:
Refactor sandbox generation script to conditionally use local registry

- Updated the generate.ts script to check for the --local-registry option and execute the appropriate action based on its value, improving flexibility in the sandbox generation process.
